src/v2i-hub/TelematicBridgePlugin/src/TelematicUnit.cpp

```diff
@@ -0,0 +1,314 @@
+#include "TelematicUnit.h"
+
+namespace TelematicBridge
+{
+    void TelematicUnit::connect(const string &natsURL)
+    {
+        auto s = natsConnection_ConnectTo(&_conn, natsURL.c_str());
+        PLOG(logINFO) << "NATS connection returned: " << natsStatus_GetText(s);
+        if (s == NATS_OK)
+        {
+            registerUnitRequestor();
+        }
+        else
+        {
+            throw TelematicBridgeException(natsStatus_GetText(s));
+        }
+    }
+
+    void TelematicUnit::registerUnitRequestor()
+    {
+        // Reset registration status
+        bool isRegistered = false;
+
+        while (!isRegistered)
+        {
+            PLOG(logDEBUG2) << "Inside register unit requestor";
+            natsMsg *reply = nullptr;
+            string payload = "{\"unit_id\":\"" + _unit.unitId + "\"}";
+            auto s = natsConnection_RequestString(&reply, _conn, REGISTER_UNIT_TOPIC, payload.c_str(), TIME_OUT);
+            if (s == NATS_OK)
+            {
+                auto replyStr = natsMsg_GetData(reply);
+                PLOG(logINFO) << "Received registered reply: " << replyStr;
+
+                // Unit is registered when server responds with event information (location, testing_type, event_name)
+                isRegistered = validateRegisterStatus(replyStr);
+                natsMsg_Destroy(reply);
+            }
+            else
+            {
+                throw TelematicBridgeException(natsStatus_GetText(s));
```

Comment:
Why do this while loop and throw and exception. If the registration fails you throw an exception so the loop is exited. If you are successful the loop is exited. Is there any condition in which you will retry. Only if NAT_OK and the validateRegisterStatus fails? In which situation would this occur.
\

Reply:
Only if we get an invalid registration reply json from telematics?

Reply:
The server can return response but it does not have event information. It throw error when NATS_OK check fails. This usually because the server is down.

Reply:
Is the response without event information also a bug or error behavior or would reattempts ever fix that. Also do we want to limit connection attempts on invalid responses?

Reply:
It is neither of them. It requires user attention to double check the event and its assigned unit.

Reply:
I think we want to wait for the valid response, otherwise there is no point to keep this plugin running.

Reply:
So if user input is required to make a valid response, then we should indicate to the user that something is wrong instead of silently indefinitely and continuously retrying a connection that has an issue that a user must resolve. 

To me either we do a single register attempt and fail or succeed and notify the user on fails, or we have a loop that attempts the connection some finite amount of times before failing. To me reattempts on connections should ideally be used for connection issues that do not require user input to be resolved, like if the server is restarting or if there are network temporary issues.

Reply:
If the concern here is uptime of the resulting plugin, maybe this registration logic needs to be more complex and have some kind of failure recovery logic but this does not seem to me to satisfy that requirement

Reply:
@paulbourelly999  Addressed https://github.com/usdot-fhwa-OPS/V2X-Hub/pull/567/commits/040a99b1b4e251b04731c572cbc2da450bd0bcd0